### PR TITLE
One-shot User Guide updates

### PR DESCRIPTION
**Pull request by shantanuchhabra**
Updated user guide based on toolkit updates and added a troubleshooting section with some known issues we're investigating.

### CODE REVIEW

File: userguide/one_shot_object_detection/README.md [hunk 1]
```diff
@@ -62,18 +61,59 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
 
 **Stage 2: Train the Object Detector model**
 
 To learn about the training implementation details, refer to the [Object Detector: How it Works](https://apple.github.io/turicreate/docs/userguide/object_detection/how-it-works.html) chapter.
 
+## Troubleshooting
+
+##### My starter image can appear upside down as well as rotated 90 degrees in the wild. Is there anything I can do to improve the quality of the model?
+
+To ensure a higher quality model, instead of providing *one* starter image, feel free to provide *more than one* starter images. As of Turi Create 5.8, for every starter image, we rotate the starter image by up to 20 degrees. As a result, to detect both inverted and erect projections of the same starter image, you can provide *2* starter images with the same label. Here is an example to rotate a Stop Sign starter image by 90 degrees, 180 degrees, and 270 degrees, and construct an SFrame with four starter images:
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: User guide should ideally not mention version numbers.

File: userguide/one_shot_object_detection/README.md [hunk 2]
```diff
@@ -62,18 +61,59 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
 
-To begin training, a set of synthetic training data needs to be generated using the provided starter image(s). 
+To begin training, the provided starter image(s) are used to generate a set of synthetic training data. 
 The synthetic images are generated by applying yaw, pitch, and roll rotations on a set of background images. If no background images are provided, a set of default background images are automatically downloaded and used.
 
 **Stage 2: Train the Object Detector model**
 
 To learn about the training implementation details, refer to the [Object Detector: How it Works](https://apple.github.io/turicreate/docs/userguide/object_detection/how-it-works.html) chapter.
 
+## Troubleshooting
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: It's usually not good to have a section and start with the subsection immediately.

File: userguide/one_shot_object_detection/README.md [hunk 3]
```diff
@@ -62,18 +61,59 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
+rotated_90 = pil_to_tc(pil_image.rotate(90, expand=True))
+rotated_180 = pil_to_tc(pil_image.rotate(180, expand=True))
+rotated_270 = pil_to_tc(pil_image.rotate(270, expand=True))
+starter_images = tc.SFrame({
+    'image': [rotated_0, rotated_90, rotated_180, rotated_270],
+    'label': ['stop_sign', 'stop_sign', 'stop_sign', 'stop_sign']
+    })
+```
+
+##### While training with the default backgrounds, my starter image does not appear in most of the images in the synthetic data.
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: A user guide is meant to be permanent content. Please file known issues on GitHub and document work arounds there.

File: userguide/one_shot_object_detection/README.md [hunk 4]
```diff
@@ -62,18 +61,59 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
+
+This is a known issue that we're trying to fix. As a workaround, please try resizing your starter image (while preserving the aspect ratio of the starter image) to get it as close to `500x500` as possible. You can leverage the following API provided by `turicreate.Image`:
+
+```python
+image = turicreate.Image('starter.png')
+turicreate.image_analysis.resize(image, width=image.width/2, height=image.height/2, channels=image.channels)
+```
+
+##### Starter image has the same (x,y) coordinates for the center of the starter image in the augmented dataset.
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: Same as above.

File: userguide/one_shot_object_detection/README.md [hunk 5]
```diff
@@ -84,16 +124,16 @@ You can preview the synthetic images the toolkit generates by calling the `previ
 
 ```python
 synthetic_images = \
-tc.one_shot_object_detector.util.preview_synthetic_training_data(starter_images, 'label')
+tc.one_shot_object_detector.util._augmentation.preview_synthetic_training_data(starter_images, 'label')
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: Do not document "_" functions in the UG. Please remove this section if it's not ready.

File: userguide/one_shot_object_detection/README.md [hunk 6]
```diff
@@ -112,5 +152,5 @@ To preview the synthetic images generated with your custom background images:
 
 ```python
 synthetic_images = \
-tc.one_shot_object_detector.util.preview_synthetic_training_data(starter_images, 'label', my_backgrounds)
+tc.one_shot_object_detector.util._augmentation.preview_synthetic_training_data(starter_images, 'label', my_backgrounds)
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: Same as above.

File: userguide/one_shot_object_detection/README.md [hunk 7]
```diff
@@ -62,18 +61,59 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
+    FORMAT_RAW = 2
+    return tc.Image(_image_data=image.tobytes(),
+                    _width=image.shape[1],
+                    _height=image.shape[0],
+                    _channels=image.shape[2],
+                    _format_enum=FORMAT_RAW,
+                    _image_data_size=image.size)
+
+import turicreate as tc
+image_path = ... set image path here ...
```
Comment (srikris) [inline userguide/one_shot_object_detection/README.md]: Please make this code more readable and copy passable. Use nice function names, variable names, and potentially refactor to make it nicer.

File: userguide/one_shot_object_detection/README.md [hunk 8]
```diff
@@ -89,11 +130,11 @@ tc.one_shot_object_detector.util.preview_synthetic_training_data(starter_images,
 synthetic_images.explore()
 ```
 
-![Explore synthetic images](images/synthetic_images_explore.jpg)
+![Explore synthetic images](images/synthetic_images_explore.png)
```
Comment (fareeha) [inline userguide/one_shot_object_detection/README.md]: Is there a reason for changing the file format from `jpg` to a heavier format `png`?

If the image files have been changed, I would recommend compressing them. For example the image below is 2.95MB now vs 916KB before.

Comment (shantanuchhabra) [reply userguide/one_shot_object_detection/README.md]: Done.

File: userguide/one_shot_object_detection/README.md [hunk 9]
```diff
@@ -1,41 +1,40 @@
...
 
 # Load test images
-test_data = tc.SFrame({'image':[tc.Image('stop_sign_photo1.png'), 
-                                tc.Image('stop_sign_photo2.png')],
-                        'label':['stop_sign', 'stop_sign']})
+test_images = tc.SFrame({'image':[tc.Image('stop_sign_test1.jpg'), 
+                                  tc.Image('stop_sign_test2.jpg')]})
 
-# Create a model. This step may take several hours                                      
+# Create a model. This step will take about an hour
```
Comment (TobyRoseman) [inline userguide/one_shot_object_detection/README.md]: Is this on CPU or GPU? I'm guessing there's a big difference.

File: userguide/one_shot_object_detection/README.md [hunk 10]
```diff
@@ -1,41 +1,40 @@
...
 ## Introductory Example
 
 In this example, our goal is to **localize** instances of **stop signs** in images. To train a model for this application, we supply a single image of a stop sign as a starter image.
 
-Given a starter image (*left*), a one-shot detector trained on this starter image will produce instance predictions on a test image (*right*) which may look like this:
+Given a starter image (*left*), a one-shot object detector trained on this starter image will produce instance predictions on a test image (*right*) which may look like this:
 
 ![Stop Sign vector and prediction images](images/stop_sign_sample.jpg) 
 
+Download the starter image to get started from [here](images/stop_sign_starter.png).
```
Comment (TobyRoseman) [inline userguide/one_shot_object_detection/README.md]: I think it pretty bad form to hyper link the word `here`. I think this is much better
```
Download the [stop_sign_starter.png](images/stop_sign_starter.png) image. Also Download [stop_sing_photo1.png](images/stop_sign_test1.jpg) and [stop_sing_photo2.png](images/stop_sign_test2.jpg)

File: userguide/one_shot_object_detection/README.md [hunk 11]
```diff
@@ -62,18 +61,60 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
 
 ## How it works
 
-In this section, we will go into detail on what happens under the hood of the `create` method. An OSOD model is trained in two steps: First, we use the starter images to generate synthetic data and annotations, and second, we train an object detector.
+In this section, we will go into detail on what happens under the hood of the `create` method. A One-Shot Object Detector model is trained in two steps: First, we use the starter images to generate synthetic data and annotations, and second, we train a model on the generated synthetic data.
```
Comment (TobyRoseman) [inline userguide/one_shot_object_detection/README.md]: The second sentence that starts on this line is actually three sentences. I would re-work as follows:
```
A One-Shot Object Detector model is trained in two steps. First, the starter image(s) is used to generate synthetic data and annotations. Second, we train an object detector model on the generated synthetic data.
```

File: userguide/one_shot_object_detection/README.md [hunk 12]
```diff
@@ -62,18 +61,60 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
+tc_image = tc.Image(image_path)
+pil_image = Image.open(image_path)
+images = [tc_image]
+for rotation_angle in [90, 180, 270]:
+    rotated = pil_to_tc(pil_image.rotate(rotation_angle, expand=True))
+    images.append(rotated)
+
+starter_images = tc.SFrame({
+    'image': images,
+    'label': ['stop_sign', 'stop_sign', 'stop_sign', 'stop_sign']
```
Comment (TobyRoseman) [inline userguide/one_shot_object_detection/README.md]: I would change this line to: 
```python
'label': ['my_label'] * 4        # set your labels here.
```

File: userguide/one_shot_object_detection/README.md [hunk 13]
```diff
@@ -62,18 +61,60 @@ To learn more about deploying One-Shot to CoreML details, refer to the [Object D
...
+})
+```
 
+##### While training with the default backgrounds, my starter image does not appear in most of the images in the synthetic data.
+
+This is a known issue ([#2341](https://github.com/apple/turicreate/issues/2341)) that we're trying to fix. As a workaround, please try resizing your starter image (while preserving the aspect ratio of the starter image) to get it as close to `500x500` (for the default backgrounds) as possible. You can leverage the following API provided by `turicreate.Image`:
+
+```python
+image = turicreate.Image('starter.png')
+turicreate.image_analysis.resize(image, width=image.width/2, height=image.height/2, channels=image.channels)
```
Comment (TobyRoseman) [inline userguide/one_shot_object_detection/README.md]: Is it cool if `image.width` or `image.height` is odd and we're not in python 2? In other words what happens if the values we pass to `resize` for `width` or `height` is a non-whole number float? Maybe change `/` -> '//'.

File: userguide/one_shot_object_detection/README.md [hunk 14]
```diff
@@ -1,41 +1,38 @@
...
 # Load the starter images
-starter_images = tc.SFrame({'image':[tc.Image('stop_sign.png')],
+starter_images = tc.SFrame({'image':[tc.Image('stop_sign_starter.png')],
                    'label':['stop_sign']})
 
 # Load test images
-test_data = tc.SFrame({'image':[tc.Image('stop_sign_photo1.png'), 
-                                tc.Image('stop_sign_photo2.png')],
-                        'label':['stop_sign', 'stop_sign']})
+test_images = tc.SFrame({'image':[tc.Image('stop_sign_test1.jpg'),
```
Comment (shreyajain17) [inline userguide/one_shot_object_detection/README.md]: It should be `test_data` I think

**Pull request merged**